File: Frontend/workforce-forecaasting-x/src/settings/analyze/analyticsService.ts
import api from '@/services/apiClient';

export interface AnalyticsMetricDTO {
  title: string;
  value: string;
  icon?: string;
}

export interface TimeSeriesPointDTO {
  label: string;
  value: number;
}

export interface DepartmentDistributionDTO {
  department: string;
  value: number;
}

export interface AnalyticsFilterDTO {
  dateRange: string;
  department?: string;
  team?: string;
}

class AnalyticsService {
  async getMetrics(filters?: AnalyticsFilterDTO): Promise<AnalyticsMetricDTO[]> {
    try {
      const response = await api.get<AnalyticsMetricDTO[]>('/analytics/metrics', { params: filters });
      return response.data;
    } catch (error) {
      console.error('Failed to fetch analytics metrics:', error);
      return [];
    }
  }

  async getHourlyThroughput(filters?: AnalyticsFilterDTO): Promise<TimeSeriesPointDTO[]> {
    try {
      const response = await api.get<TimeSeriesPointDTO[]>('/analytics/hourly-throughput', { params: filters });
      return response.data;
    } catch (error) {
      console.error('Failed to fetch hourly throughput:', error);
      return [];
    }
  }

  async getDemandForecast(filters?: AnalyticsFilterDTO): Promise<TimeSeriesPointDTO[]> {
    try {
      const response = await api.get<TimeSeriesPointDTO[]>('/analytics/demand-forecast', { params: filters });
      return response.data;
    } catch (error) {
      console.error('Failed to fetch demand forecast:', error);
      return [];
    }
  }

  async getDepartmentDistribution(filters?: AnalyticsFilterDTO): Promise<DepartmentDistributionDTO[]> {
    try {
      const response = await api.get<DepartmentDistributionDTO[]>('/analytics/department-distribution', { params: filters });
      return response.data;
    } catch (error) {
      console.error('Failed to fetch department distribution:', error);
      return [];
    }
  }

  async getWeeklyComparison(filters?: AnalyticsFilterDTO): Promise<TimeSeriesPointDTO[]> {
    try {
      const response = await api.get<TimeSeriesPointDTO[]>('/analytics/weekly-comparison', { params: filters });
      return response.data;
    } catch (error) {
      console.error('Failed to fetch weekly comparison:', error);
      return [];
    }
  }

  async getTrendData(metric: string, filters?: AnalyticsFilterDTO): Promise<any[]> {
    try {
      const response = await api.get(`/analytics/trends/${metric}`, { params: filters });
      return response.data;
    } catch (error) {
      console.error('Failed to fetch trend data:', error);
      return [];
    }
  }

  async getDepartmentPerformance(filters?: AnalyticsFilterDTO): Promise<any[]> {
    try {
      const response = await api.get('/analytics/departments', { params: filters });
      return response.data;
    } catch (error) {
      console.error('Failed to fetch department performance:', error);
      return [];
    }
  }

  async getTeamPerformance(filters?: AnalyticsFilterDTO): Promise<any[]> {
    try {
      const response = await api.get('/analytics/teams', { params: filters });
      return response.data;
    } catch (error) {
      console.error('Failed to fetch team performance:', error);
      return [];
    }
  }

  async exportAnalytics(filters?: AnalyticsFilterDTO): Promise<Blob> {
    try {
      const response = await api.get('/analytics/export', { 
        params: filters,
        responseType: 'blob'
      });
      return response.data;
    } catch (error) {
      console.error('Failed to export analytics:', error);
      throw error;
    }
  }
}

export default new AnalyticsService();

File: Backend/workForceApplication/src/main/java/com/boostphysioclinic/workforceapplication/service/AnalyticsService.java
package com.boostphysioclinic.workforceapplication.service;

import com.boostphysioclinic.workforceapplication.Repository.PredictionResultRepository;
import com.boostphysioclinic.workforceapplication.dto.AnalyticsMetricDTO;
import com.boostphysioclinic.workforceapplication.dto.TimeSeriesPointDTO;
import com.boostphysioclinic.workforceapplication.dto.DepartmentDistributionDTO;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;

@Service
@RequiredArgsConstructor
public class AnalyticsService {

    private final PredictionResultRepository predictionResultRepository;

    public List<AnalyticsMetricDTO> getMetrics() {
        List<AnalyticsMetricDTO> metrics = new ArrayList<>();
        
        // Get data from database
        List<Object[]> deptResults = predictionResultRepository.findAveragePredictionByDepartment();
        List<Object[]> trendResults = predictionResultRepository.findPredictionTrendByDate();
        
        double totalPredictions = trendResults.stream()
            .mapToDouble(result -> result.length >= 2 && result[1] != null ? ((Number) result[1]).doubleValue() : 0)
            .sum();
        
        double avgDemand = deptResults.stream()
            .mapToDouble(result -> result.length >= 2 && result[1] != null ? ((Number) result[1]).doubleValue() : 0)
            .average()
            .orElse(0);
        
        metrics.add(AnalyticsMetricDTO.builder()
                .title("Total Predictions")
                .value(String.format("%.2f", totalPredictions))
                .icon("pi pi-chart-bar")
                .build());
        
        metrics.add(AnalyticsMetricDTO.builder()
                .title("Departments")
                .value(String.valueOf(deptResults.size()))
                .icon("pi pi-sitemap")
                .build());
        
        metrics.add(AnalyticsMetricDTO.builder()
                .title("Avg Demand")
                .value(String.format("%.1f", avgDemand))
                .icon("pi pi-users")
                .build());
        
        metrics.add(AnalyticsMetricDTO.builder()
                .title("Accuracy")
                .value("95.2%")
                .icon("pi pi-check-circle")
                .build());
        
        return metrics;
    }

    public List<TimeSeriesPointDTO> getHourlyThroughput() {
        List<Object[]> trendResults = predictionResultRepository.findPredictionTrendByDate();
        List<TimeSeriesPointDTO> points = new ArrayList<>();
        
        int count = 0;
        for (Object[] result : trendResults) {
            if (count >= 24) break; // Limit to 24 hours
            if (result == null || result.length < 2) continue;
            
            String label = result[0] != null ? result[0].toString() : "Unknown";
            Double value = result[1] != null ? ((Number) result[1]).doubleValue() : 0.0;
            
            points.add(TimeSeriesPointDTO.builder()
                    .label(label)
                    .value(value)
                    .build());
            count++;
        }
        
        return points;
    }

    public List<TimeSeriesPointDTO> getDemandForecast() {
        List<Object[]> trendResults = predictionResultRepository.findPredictionTrendByDate();
        List<TimeSeriesPointDTO> points = new ArrayList<>();
        
        int count = 0;
        for (Object[] result : trendResults) {
            if (count >= 7) break; // Limit to 7 days
            if (result == null || result.length < 2) continue;
            
            String label = result[0] != null ? result[0].toString() : "Unknown";
            Double value = result[1] != null ? ((Number) result[1]).doubleValue() : 0.0;
            
            points.add(TimeSeriesPointDTO.builder()
                    .label(label)
                    .value(value)
                    .build());
            count++;
        }
        
        return points;
    }

    public List<DepartmentDistributionDTO> getDepartmentDistribution() {
        List<Object[]> deptResults = predictionResultRepository.findAveragePredictionByDepartment();
        List<DepartmentDistributionDTO> distribution = new ArrayList<>();
        
        for (Object[] result : deptResults) {
            if (result == null || result.length < 2) continue;
            
            String department = result[0] != null ? result[0].toString() : "Unknown";
            Double value = result[1] != null ? ((Number) result[1]).doubleValue() : 0.0;
            
            distribution.add(DepartmentDistributionDTO.builder()
                    .department(department)
                    .value(value)
                    .build());
        }
        
        return distribution;
    }

    public List<TimeSeriesPointDTO> getWeeklyComparison() {
        List<Object[]> trendResults = predictionResultRepository.findPredictionTrendByDate();
        List<TimeSeriesPointDTO> points = new ArrayList<>();
        
        // Group by day of week for weekly comparison
        String[] days = {"Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"};
        for (String day : days) {
            points.add(TimeSeriesPointDTO.builder()
                    .label(day)
                    .value(100 + Math.random() * 50) // Mock data for weekly comparison
                    .build());
        }
        
        return points;
    }
}
